### PR TITLE
Add skipBatch Capacity for IndexedVectorizedOapRecordReader

### PR DESCRIPTION
**Pull request by LuciferYang**
## What changes were proposed in this pull request?

SubTask of #825 

In the past `IndexedVectorizedOapRecordReader.readBatch` method will real assembly batch data to `ColumnVector` even if we not needed this batch, for index read with large RowGroup there may some extra assembly cost, in this patch we use skipBatch capacity  of `SkippableVectorizedColumnReader`  to optimization this problem.

## How was this patch tested?
exists `VectorizedDataSuite` cover this features.
add  `IndexedVectorizedOapReaderSuite` to test skip behavior
mvn test pass
mvn test -Pspark-2.2 pass
mvn test -Pspark-2.3 pass

### CODE REVIEW

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 1]
```diff
@@ -102,7 +118,23 @@ public boolean nextBatch() throws IOException {
...
-      return super.nextBatch() && filterRowsWithIndex();
+
+      if (rowsReturned >= totalRowCount) return false;
+
+      checkEndOfRowGroup();
+
+      IntList ids = idsMap.remove(currentPageNumber++);
+
+      // when we do this while loop there must be remainder pageNumber->ids in idsMap，
+      // it guarantee by rowsReturned value and checkEndOfRowGroup method.
```
Comment (linhong-intel) [inline src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: Can we do something like below.
```
while(!idsMap.isEmpty()) {
  ids = idsMap.remove(currentPageNumber++);
  if (!ids.isEmpty()) break;
  skipBatchInternal();
}
nextBatchInternal();
filterRowsWIthIndex(ids);
```
Then we may not need depend `rowsReturned` and `checkEndOfRowGroup` to guarantee the `idsMap`

File: src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java [hunk 2]
```diff
@@ -286,9 +286,18 @@ public void enableReturningBatches() {
      * From VectorizedParquetRecordReader, no change.
      */
     public boolean nextBatch() throws IOException {
-      columnarBatch.reset();
       if (rowsReturned >= totalRowCount) return false;
```
Comment (linhong-intel) [inline src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java]: is it ok to leave the last columnarBatch "not reset"?

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java (hunk 1 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: let me test it

File: src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java (hunk 2 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java]: if return false, this columnarBatch never be used again, the state is not important.

**Comment by linhong-intel** [comment]
lgtm

**Comment by LuciferYang** [comment]
cc @gczsjdy

**Comment by LuciferYang** [comment]
@gczsjdy  help me review this pr?

File: src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+   * | no index hit batch | --> call skipBatchInternal to skip
+   * ----------------------
+   * | hit 16000 batch    | --> call nextBatchInternal() && filterRowsWithIndex(ids)
+   * ----------------------
+   * | no index hit batch | --> discard directly because of idsMap.isEmpty()
+   * ----------------------
+   * Now this case a little trick, we use spy model to mock IndexedVectorizedOapReader object,
+   * it will call real method and we only verify call times of Key methods.
+   */
+  test("IndexedVectorizedOapRecordReader - readBatch") {
```
Comment (gczsjdy) [inline src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala]: :nit nextBatch

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 4]
```diff
@@ -115,37 +147,30 @@ protected void checkEndOfRowGroup() throws IOException {
...
+        // we can do same operation use markFiltered as follow code:
+        // int current = 0;
+        // for (Integer target : ids)
+        // { while (current < target){
+        // columnarBatch.markFiltered(current);
+        // current++; }
+        // current++; }
+        // current++;
+        // while (current < numBatched){
+        // columnarBatch.markFiltered(current); current++; }
```
Comment (gczsjdy) [inline src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: Why keep the commented code?

File: src/main/spark2.1/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java [hunk 5]
```diff
@@ -275,9 +275,18 @@ public void enableReturningBatches() {
      * From VectorizedParquetRecordReader, no change.
      */
     public boolean nextBatch() throws IOException {
-      columnarBatch.reset();
       if (rowsReturned >= totalRowCount) return false;
       checkEndOfRowGroup();
+      nextBatchInternal();
```
Comment (gczsjdy) [inline src/main/spark2.1/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java]: Actually just switched this from L278 -> 280? Confused...

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 6]
```diff
@@ -94,6 +95,21 @@ public Object getCurrentValue() throws IOException, InterruptedException {
...
+     * | no index hit batch | --> call skipBatchInternal to skip
+     * ----------------------
+     * | index hit batch    | --> call nextBatchInternal() && filterRowsWithIndex(ids)
+     * ----------------------
+     * | no index hit batch | --> call skipBatchInternal to skip
+     * ----------------------
+     * | index hit batch    | --> call nextBatchInternal() && filterRowsWithIndex(ids)
+     * ----------------------
+     * | no index hit batch | --> discard directly because of idsMap.isEmpty()
+     * ----------------------
```
Comment (gczsjdy) [inline src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: The diagram is good, but is it better to indicate each steps of the progress?
Like 
```
1. Check if idsMap is empty, if ...
2. ...
```

File: src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        // nextKeyValue call 3 times, result are [true , true , false]
+        Mockito.verify(reader, Mockito.times(3)).nextKeyValue()
+        // getCurrentValue call 2 times, because of we need only 2 batch of 5
+        Mockito.verify(reader, Mockito.times(2)).getCurrentValue
+        // nextBatch call 3 times, result are [true , true , false]
+        Mockito.verify(reader, Mockito.times(3)).nextBatch()
+        // nextBatchInternal call 2 times, because of we need only 2 batch of 5
+        Mockito.verify(reader, Mockito.times(2)).nextBatchInternal()
+        // skipBatchInternal call 2 times, last batch discard directly because of idsMap.isEmpty()
+        Mockito.verify(reader, Mockito.times(2)).skipBatchInternal()
```
Comment (gczsjdy) [inline src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala]: Like this : )

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 8]
```diff
@@ -102,7 +118,23 @@ public boolean nextBatch() throws IOException {
       if (idsMap.isEmpty()) {
         rowsReturned = totalCountLoadedSoFar;
       }
-      return super.nextBatch() && filterRowsWithIndex();
+
+      if (rowsReturned >= totalRowCount) return false;
+
+      checkEndOfRowGroup();
+
+      IntList ids = idsMap.remove(currentPageNumber++);
```
Comment (gczsjdy) [inline src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: idsMap is columnarBatchId -> selected rowIdList in this CB?

File: src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+      val df = spark.createDataFrame(data)
+      val schema = df.schema.json
+      df.repartition(1).write.parquet(dir.getCanonicalPath)
+      val file = SpecificParquetRecordReaderBase.listDirectory(dir).get(0)
+      val path = new Path(file.asInstanceOf[String])
+
+      configuration.set(ParquetReadSupportWrapper.SPARK_ROW_REQUESTED_SCHEMA, schema)
+
+      // TODO refactor IndexedVectorizedOapReader to use Mockito.mock
+      val reader = Mockito.spy(new TestIndexedVectorizedOapReader(path, configuration, rowIds))
```
Comment (gczsjdy) [inline src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala]: I think we can new `IndexedVectorizedOapReader` here, can't we?

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java (hunk 8 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]:
```
// pageNumber -> rowIdsList, use to decide：
    // 1. Is this pageNumber has data to read ?
    // 2. Use rowIdsList to mark which row need read.
    private Map<Integer, IntList> idsMap = Maps.newHashMap();
```

File: src/main/spark2.1/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java (hunk 5 above)
Comment (LuciferYang) [reply src/main/spark2.1/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java]: For extract nextBatchInternal method use by VectorizedOapRecordReader and IndexedVectorizedOapRecordReader, don't want to do `columnarBatch.reset()` two times and I think it call ` call to columnarBatch. reset ()` should after `if (rowsReturned >= totalRowCount) return false` judgment, if `return false` directly we can save some operations.

File: src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala (hunk 3 above)
Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala]: ok~

File: src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala (hunk 9 above)
Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala]: No, if we use `new  IndexedVectorizedOapReader`, we can't verify call times by Mockito

**Comment by LuciferYang** [comment]
@gczsjdy already add some new comments to explain sample data read process.

File: src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala (hunk 7 above)
Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/parquet/hadoop/IndexedVectorizedOapReaderSuite.scala]: yeah~

**Pull request merged**